src/eigenvalue.cpp

```diff
@@ -0,0 +1,654 @@
+    // to use all three estimators.
+    i = -1;
+  }
+
+  if (i == -1) {
+    // Use three estimators as derived in the paper by Urbatsch
+
+    // Initialize variables
+    double g = 0.0;
+    std::array<double, 3> S {};
+
+    for (int l = 0; l < 3; ++l) {
+      // Permutations of estimates
+      switch (l) {
+      case 0:
+        // i = collision, j = absorption, k = tracklength
+        i = 0;
+        j = 1;
+        k = 2;
+        break;
+      case 1:
+        // i = absortion, j = tracklength, k = collision
+        i = 1;
+        j = 2;
+        k = 0;
+        break;
+      case 2:
+        // i = tracklength, j = collision, k = absorption
+        i = 2;
+        j = 0;
+        k = 1;
+        break;
+      }
+
+      // Calculate weighting
+      double f = cov(j, j) * (cov(k, k) - cov(i, k)) - cov(k, k) * cov(i, j) +
+        cov(j, k) * (cov(i, j) + cov(i, k) - cov(j, k));
+
+      // Add to S sums for variance of combined estimate
+      S[0] += f * cov(0, l);
+      S[1] += (cov(j, j) + cov(k, k) - 2.0 * cov(j, k)) * kv[l] * kv[l];
+      S[2] += (cov(k, k) + cov(i, j) - cov(j, k) - cov(i, k)) * kv[l] * kv[j];
+
+      // Add to sum for combined k-effective
+      k_combined[0] += f * kv[l];
+      g += f;
+    }
+
+    // Complete calculations of S sums
+    for (auto& S_i : S) {
+      S_i *= (n - 1);
+    }
+    S[0] *= (n - 1)*(n - 1);
+
+    // Calculate combined estimate of k-effective
+    k_combined[0] /= g;
+
+    // Calculate standard deviation of combined estimate
+    g *= (n - 1)*(n - 1);
+    k_combined[1] = std::sqrt(S[0] / (g*n*(n - 3)) *
+      (1 + n*((S[1] - 2*S[2]) / g)));
+
+  } else {
+    // Use only two estimators
+    // These equations are derived analogously to that done in the paper by
+    // Urbatsch, but are simpler than for the three estimators case since the
+    // block matrices of the three estimator equations reduces to scalars here
+
+    // Store the commonly used term
+    double f = kv[i] - kv[j];
+    double g = cov(i, i) + cov(j, j) - 2.0*cov(i, j);
+
+    // Calculate combined estimate of k-effective
+    k_combined[0] = kv[i] - (cov(i, i) - cov(i, j)) / g * f;
+
+    // Calculate standard deviation of combined estimate
+    k_combined[1] = (cov(i, i)*cov(j, j) - cov(i, j)*cov(i, j)) *
+          (g + n*f*f) / (n*(n - 2)*g*g);
+    k_combined[1] = std::sqrt(k_combined[1]);
+
+  }
+  return 0;
+}
+
+void shannon_entropy()
+{
+  // Get source weight in each mesh bin
+  bool sites_outside;
+  xt::xtensor<double, 1> p = simulation::entropy_mesh->count_sites(
+    simulation::fission_bank, &sites_outside);
+
+  // display warning message if there were sites outside entropy box
+  if (sites_outside) {
+    if (mpi::master) warning("Fission source site(s) outside of entropy box.");
+  }
+
+  // sum values to obtain shannon entropy
+  if (mpi::master) {
+    // Normalize to total weight of bank sites
+    p /= xt::sum(p);
+
+    double H = 0.0;
```

Comment:
Nit: Is `H` a term related to shannon entropy?  Or the result? Might make it more readable if there was a comment here, or a clarifying comment where `H` gets pushed into the vector on line 564.